### PR TITLE
feat: add support for empty bar style to `Sparkline`

### LINKED ISSUE #1325: Sparkline: Add support for empty bar style
## Problem

The `Sparkline` widget takes a `&[64]` and shows a blank vertical bar for `0` values; it is desirable to be able to show `0` samples with an alternative styling to highlight them.

Today:

image

Desired:

image

## Solution

PR: https://github.com/ratatui-org/ratatui/pull/1326

Note that this solution is purely additive, there is no change to behaviour or rendering of existing `Sparkline` users.

## Alternatives

As well as controlling the style of empty bars, it may also be desirable to control the symbol as well, though I have need of that feature today.

## Additional context

In [Trippy](https://github.com/fujiapple852/trippy), we use the `Sparkline` to show a history of samples and missing samples (i.e. lost packets) are represented by zero and are therefore rendered (or rather, not rendered) as a blank vertical bar.  By styling zero values in an alternative way this can be highlighted to the user.

See https://github.com/fujiapple852/trippy/issues/1247

### PR DESCRIPTION
**Pull request by fujiapple852**
Closes #1325 

The PR adds the ability to:

- distingiush between empty bars and bars with a value of 0
- provide custom styling for empty bars
- provide custom styling for individual bars
- inverts the rendering algorithm to be item first

### CODE REVIEW

**Comment by fujiapple852** [comment]
The line missing from code coverage is an existing gap.  To fix that you'd need a test case with a width that has a larger vertical size.  Likely not worth doing.

File: src/widgets/sparkline.rs
Comment (joshka) [inline src/widgets/sparkline.rs:188]: Looking at this line in particular, it sort of suggests an idea that this could be more flexible to implement in a way that each data item is Option<u64> instead of treating 0 as a special case. This would allow the Sparkline to indicate both 0 data and missing data points on the same graph. This might be more difficult to work out a good migration strategy for however - likely adding a new method to support this (unless there's a really good way to make the method accept both datasets - I can't see an obvious one though)

**Review by joshka — COMMENTED** [review]
Overall I'm ok with this idea (it's pretty reasonable), but I'd like to explore the idea of adding a custom symbol for this a bit before committing to this.

I'll copy this discussion over to the issue though rather than here to avoid bifurcating the ideas...

File: src/widgets/sparkline.rs
Comment (joshka) [inline src/widgets/sparkline.rs:156]: Non-blocking idea. This seems like it might be easy to extend to style individual data points in the future. Do you see any problem with this data type that would prevent us from making that sort of change (I don't but it's worth getting a second opinion on this). Are there any other obvious changes that we'd want to do here?

File: src/symbols.rs
Comment (joshka) [inline src/symbols.rs:121]: Perhaps just some pub consts instead of an Enum for these?

**Review by joshka — COMMENTED** [review]
This looks good. It's (mostly) backwards compatible. The breaking change is when someone is calls e.g. `.data(some_value.into())` as the compiler cannot infer the type. We've got a few of these type changes in the breaking changes doc. Can you add one for this change too.

This means it will be part of 0.29.0, mainly to avoid breaking apps that are not installed using the `--locked` arg to `cargo install`

File: src/widgets/sparkline.rs
Comment (fujiapple852) [reply src/widgets/sparkline.rs]: After some experimentation, i've moved `EmptyBarSymbol` into the `sparkline` module and added a `Custom(String)` variant.  

I tried to use `char` as I feel we should restrict this, but given unicode formatting we can't assume much about the "width" anyway.

File: src/widgets/sparkline.rs
Comment (fujiapple852) [reply src/widgets/sparkline.rs:1]: Added.  I tidied up the main doc a bit as well as there were some things missing.

File: src/widgets/sparkline.rs
Comment (fujiapple852) [reply src/widgets/sparkline.rs]: Added.  

I've split `SparklineData` into `SparkData(Vec<SparkItem>)`.  Happy to bike shed on names.

`Sparks` / `Spark`?
`SparkLineData` / `SparkLineItem`?
`Data` / `Item`?

File: src/widgets/sparkline.rs
Comment (fujiapple852) [reply src/widgets/sparkline.rs]: I've added optional per-item styling.  

The styles cascade, the `Sparkline` stying will be used by default but can be overridden per item (i.e. `SparkItem`).  Note that empty item styling remains independent.

There are some news tests to demonstrate these "rainbow" Sparklines.

File: src/symbols.rs
Comment (fujiapple852) [reply src/symbols.rs]: Now that this has been moved and includes a `Custom(String)` case I think it makes sense to remain as a enum.

**Comment by fujiapple852** [comment]
@joshka pushed a 5th commit (will squash later) with updates, seem comments in threads.  Still a WIP, but if you're happy with the latest approach i'll polish up for merging.

File: src/widgets/sparkline.rs
Comment (fujiapple852) [inline src/widgets/sparkline.rs:316]: I think most of these `From impls` can be merged, i'll try to do that.

File: src/widgets/sparkline.rs
Comment (fujiapple852) [inline src/widgets/sparkline.rs:63]: Should this be a doc comment?

File: src/widgets/sparkline.rs
Comment (fujiapple852) [reply src/widgets/sparkline.rs]: Updated to unify these

File: src/widgets/sparkline.rs
Comment (fujiapple852) [reply src/widgets/sparkline.rs]: I've managed to eliminate `SparkData`, so now the signature of data is reduced to:

```rust
    pub fn data<T>(mut self, data: T) -> Self
    where
        T: IntoIterator,
        T::Item: Into<SparkItem>
```

This should, I suspect, also reduce issues with type inference.

**Comment by fujiapple852** [comment]
@joshka i've tidied this up and rebased, I think it about ready, but let me know if you have any other feedback.

The only test failure is the semver check as the `Sparkline::data` function is no longer `const`, I don't think we can do anything about that, unless we want to keep the legacy `Sparkline::data` const function as-is and add an additional non-const function (`Sparkline::data2` or something).

Performance-wise, this version clones the input data twice vs once in the prior version.  We could, with a lot of work, remove one of these clones, but it would need a fairly significant rework of the render function (it renders row-first and mutates the data as it goes to keep track, we'd have to invert it render column-first).  Not something I feel like attempting right now.

File: src/widgets/sparkline.rs
Comment (fujiapple852) [inline src/widgets/sparkline.rs:167]: Remove reference to `SparkData`

File: src/widgets/sparkline.rs
Comment (fujiapple852) [inline src/widgets/sparkline.rs:190]: Remove reference to `SparkData`

File: src/widgets/sparkline.rs
Comment (fujiapple852) [inline src/widgets/sparkline.rs:154]: Typo

File: src/widgets/sparkline.rs
Comment (fujiapple852) [inline src/widgets/sparkline.rs:178]: Remove reference to `SparkData`

**Comment by fujiapple852** [comment]
@joshka I nerd sniped myself into rewriting the Sparkline render method to be more efficient, by avoiding the data clone that is performed today.  By doing so, the Sparkline remains net neutral in terms of allocation.

The existing implementation renders row-by-row, and for each row, item-by-item.  To make this work it has to keep track of how much height remains to be rendered for each item (as that determined the symbol to render).  It does this be cloning the data and mutating it as it renders.

The new implementation renders item-by-item, and for each item, row-by-row.  Inverting the implementation in this way means we only need to keep track of the remaining height of a single item at a time (a single u64 on the stack) and thus there is no need to clone the data, nor mutate it during the render.

The new implementation passes all the tests and seems to work when used in a real Tui app, but there is no doubt this is a much bigger and riskier change.  I've added this as a separate commit as it may make sense to leave this out from this PR and do as a separate follow up.

**Comment by orhun** [comment]
Yup, it would be great to separate the performance improvement from this PR. From your description, it doesn’t sound too bad, to be honest, and I'd be happy to review it after you follow up :)

File: src/widgets/sparkline.rs
Comment (joshka) [inline src/widgets/sparkline.rs:352]: This for loop can probably be simplified to something like:
```rust
let cols = if self.direction == RenderDirectionLeftToRight { area.cols() } else { area.cols.rev() };
for (item, col) in zip(self.data, cols) {
   ...
}
```

**Review by joshka — COMMENTED** [review]
Normally I'd agree that splitting this into a couple of smaller PRs would be good, but it's only a small amount above my threshold for review pain (usually around 200 LoC), and the changes involved are reasonably related. So I've given this a proper review as if it's going ahead in this PR rather than being split.

Grouping my review comments:
- some missing docs (empty items / style overrides)
- consider using constants instead of an enum
- individual styles should override / combine rather than replace styles set on the widget
- Some bikeshedding on names (SparkItem, bar)
- Some refactoring ideas (move the single bar rendering code into the SparkItem)
- Some questions (max, perf)
- Some missing tests (array, 

Of those, the main ones to focus on are are docs, enum, style overrides and the bikeshedding on SparkItem.

General approval on these changes
- rendering column by column instead of row by row
- only rendering the displayed columns
- adding the empty bar / styling
- adding the style per bar

File: src/widgets/sparkline.rs
Comment (fujiapple852) [reply src/widgets/sparkline.rs:300]: This is needed for cases such as `vec![Some(1_u64), None, Some(3)]` which become an iterator of `&Option<u64>`.

aside: ideally I'd like to unify the 4x `From<SparkBar>` impls, I did try and briefly before and ran into issues with overlapping impls and the orphan rule so left it as it was.

File: src/widgets/sparkline.rs
Comment (fujiapple852) [reply src/widgets/sparkline.rs:374]: I didn't give it much thought as it was this way in the original impl. 

Interestingly, if an explicit max is _not_ given, and cannot be calculated, the default is set to 1 (in both the old and new impls):

Old:

```
let max = self
    .max
    .unwrap_or_else(|| *self.data.iter().max().unwrap_or(&1));
```

New:

```
// determine the maximum height across all bars
let max_height = self
    .max
    .unwrap_or_else(|| self.data.iter().filter_map(|s| s.value).max().unwrap_or(1));
```

So the only way `max_height` could be zero is if it is explicitly set as zero by calling `Sparkline::max(u64)`.  Perhaps that API should be modified to take a `NonZeroU64` instead?

File: src/widgets/sparkline.rs
Comment (fujiapple852) [reply src/widgets/sparkline.rs:400]: `.set_style(self.style.patch(style.unwrap_or_default()))` seems to be what we want, it will combine the styles by layering the bar style over the width style.

File: src/widgets/sparkline.rs
Comment (fujiapple852) [reply src/widgets/sparkline.rs:352]: I tried with:

```rust
let cols = if self.direction == RenderDirection::LeftToRight {
    spark_area.columns()
} else {
    spark_area.columns().rev()
};
for (item, col) in self.data.iter().take(max_index).zip(cols) { 
  ...snip...
  buf[(col.x, spark_area.top() + j)]
      .set_symbol(symbol)
      .set_style(self.style.patch(style.unwrap_or_default()));
}
```

However it fails with `rev()` as:

```
the trait `DoubleEndedIterator` is not implemented for `rect::iter::Columns`
```

File: src/widgets/sparkline.rs
Comment (fujiapple852) [reply src/widgets/sparkline.rs]: Added to `BREAKING-CHANGES.md`:

```
- [v0.29.0](#v0290)
  - `Sparkline::data` is no longer declared as `const`
  - `Sparkline::data` now takes `IntoIterator<Item = SparklineBar>` instead of `&[u64]`
```

I'm unclear on what to add to the documentation based on your example, can you clarify?  Perhaps starting with an example where inference will be an issue and then showing the recommended solution.

File: src/widgets/sparkline.rs
Comment (fujiapple852) [reply src/widgets/sparkline.rs:69]: Moved back to symbols module and changed to be const values.

File: src/widgets/sparkline.rs
Comment (fujiapple852) [reply src/widgets/sparkline.rs:23]: I renamed as `AbsentValueSymbol` throughout (i.e. `absent_value_style` & `absent_value_symbol`) and updated the documentation.

File: src/widgets/sparkline.rs
Comment (fujiapple852) [reply src/widgets/sparkline.rs]: Added

File: src/widgets/sparkline.rs
Comment (fujiapple852) [reply src/widgets/sparkline.rs]: Added

File: src/widgets/sparkline.rs
Comment (fujiapple852) [reply src/widgets/sparkline.rs]: Updated the language:

```rust
/// If the data provided is a slice of `u64` or `Option<u64>`, the bars will be styled with the
/// style of the sparkline. If the data is a slice of [`SparklineBar`], the bars will be
/// styled with the style of the sparkline combined with the style provided in the [`SparklineBar`]
/// if it is set, otherwise the sparkline style will be used.
```

File: src/widgets/sparkline.rs
Comment (fujiapple852) [reply src/widgets/sparkline.rs]: I went with `SparklineBar`.

File: src/widgets/sparkline.rs
Comment (fujiapple852) [reply src/widgets/sparkline.rs:345]: I did initially try pulling some code out from the loop (but still within `Sparkline` impl) but I felt the result was worse as there is quite a lot of state to move in and out of fn calls.  I did pull `symbol_for_height` out which I think helps a bit.

File: src/widgets/sparkline.rs
Comment (fujiapple852) [reply src/widgets/sparkline.rs:561]: Added both

File: src/widgets/sparkline.rs
Comment (fujiapple852) [reply src/widgets/sparkline.rs]: Given the renderer no longer performs a clone, the new version is net natural overall in terms of performance.

File: src/widgets/sparkline.rs
Comment (fujiapple852) [reply src/widgets/sparkline.rs:250]: Added:

```rust
/// If not set, the default style of the sparkline will be used.
///
/// As well as the style of the sparkline, each [`SparklineBar`] may optinally set its own
/// style.  If set, the style of the bar will be the style of the sparkline combined with
/// the style of the bar.
```

**Comment by fujiapple852** [comment]
@orhun @joshka I've squashed the changes so now the rendering changes are part of this overall change, which has grown somewhat larger in scope than I had intended :). But I feel more comfortable with the renderer changes now given they've had an additional set of eyes on them.

I've addressed all of the review comments, I've left open a couple of threads where I need your input, one on documenting the type inference implications, and the other on the use of the `Columns` iterator.

File: src/widgets/sparkline.rs
Comment (joshka) [reply src/widgets/sparkline.rs]: I like the shade module.

File: src/widgets/sparkline.rs
Comment (joshka) [reply src/widgets/sparkline.rs:372]: This is definitely something to come back to later if needed. No need to fix for now.

File: src/widgets/sparkline.rs
Comment (joshka) [reply src/widgets/sparkline.rs:350]: Makes sense. That's addable, but not necessary for this PR.

File: src/widgets/sparkline.rs
Comment (joshka) [reply src/widgets/sparkline.rs]: The broken code is anything where you'd pass a value that can be converted into u64.
A simpler example of this is:
```rust
let value = 1u8;
Sparkline::default().data(&[value.into()]);
```
The fix is:
```rust
Sparkline::default().data(&[u64::from(value)]);
```

File: BREAKING-CHANGES.md
Comment (joshka) [inline BREAKING-CHANGES.md:15]: Can you please add the corresponding detail below (mark the section as unreleased)

File: src/widgets/sparkline.rs
Comment (joshka) [inline src/widgets/sparkline.rs:30]:
```suggestion
/// Absent values and will be rendered with the style set by [`Sparkline::absent_value_style`] and
```

File: src/widgets/sparkline.rs
Comment (joshka) [inline src/widgets/sparkline.rs:146]:
```suggestion
    /// Absent values and will be rendered with the style set by [`Sparkline::absent_value_style`]
```

File: src/widgets/sparkline.rs
Comment (joshka) [reply src/widgets/sparkline.rs:352]: Sounds good

**Review by joshka — APPROVED** [review]
LGTM with a couple of small caveats:
- A couple of small typos
- Some detail to add in the BREAKING CHANGES doc.

**Comment by joshka** [comment]
Thanks for all your hard work getting this over the line.

File: BREAKING-CHANGES.md
Comment (fujiapple852) [reply BREAKING-CHANGES.md:15]: Added:

> ## Unreleased

### `Sparkline::data` takes `IntoIterator<Item = SparklineBar>` instead of `&[u64]` and is no longer const ([#1326])

[#1326]: https://github.com/ratatui/ratatui/pull/1326

The `Sparkline::data` method has been modified to accept `IntoIterator<Item = SparklineBar>` 
instead of `&[u64]`. 

`SparklineBar` is a struct that contains an `Option<u64>` value, which represents an possible 
_absent_ value, as distinct from a `0` value. This change allows the `Sparkline` to style 
data points differently, depending on whether they are present or absent.

`SparklineBar` also contains an `Option<Style>` that will be used to apply a style the bar in 
addition to any other styling applied to the `Sparkline`.

Several `From` implementations have been added to `SparklineBar` to support existing callers who 
provide `&[u64]` and other types that can be converted to `SparklineBar`, such as `Option<u64>`.

If you encounter any type inference issues, you may need to provide an explicit type for the data
passed to `Sparkline::data`. For example, if you are passing a single value, you may need to use
`into()` to convert it to form that can be used as a `SparklineBar`:

```diff
let value = 1u8;
- Sparkline::default().data(&[value.into()]);
+ Sparkline::default().data(&[u64::from(value)]);
```

As a consequence of this change, the `data` method is no longer a `const fn`.

File: src/widgets/sparkline.rs
Comment (fujiapple852) [reply src/widgets/sparkline.rs]: Understood.  I added the example and some commentary to the `BREAKING-CHANGES.md` details.

File: BREAKING-CHANGES.md
Comment (joshka) [reply BREAKING-CHANGES.md]: perfect - thanks!

**Comment by joshka** [comment]
I'd probably suggest holding off merging this until we're close to a 0.29.0 release - mainly because it means that we can release any non-breaking stuff in the mean time.

**Review by joshka — APPROVED** [review]

**Pull request merged**
